src/Interpreters/addSubcolumnsExtraction.cpp

```diff
@@ -0,0 +1,82 @@
+#include <Core/Block.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/NestedUtils.h>
+#include <Functions/FunctionFactory.h>
+#include <Interpreters/addSubcolumnsExtraction.h>
+#include <iostream>
+
+namespace DB
+{
+
+ActionsDAG addSubcolumnsExtraction(ActionsDAG dag, const Block & header, const ContextPtr & context)
+{
+    /// First check if we need to extract anything or not.
+    bool need_to_extract_subcolumns = false;
+    for (const auto & input : dag.getInputs())
+    {
+        if (!header.has(input->result_name) && header.findSubcolumnByName(input->result_name).has_value())
+        {
+            need_to_extract_subcolumns = true;
+            break;
+        }
+    }
+
+    if (!need_to_extract_subcolumns)
+        return dag;
+
+    /// Create a new ActionsDAG that only extracts all subcolumns that are not presented in the header.
+    ActionsDAG extract_subcolumns_dag;
+
+    /// Add inputs that are presented in the header.
+    std::unordered_map<String, const ActionsDAG::Node *> input_nodes;
+    for (const auto & input : dag.getInputs())
+    {
+        if (header.has(input->result_name))
+        {
+            const auto & input_node = extract_subcolumns_dag.addInput(input->result_name, input->result_type);
+            extract_subcolumns_dag.addOrReplaceInOutputs(input_node);
+            input_nodes[input->result_name] = &input_node;
+        }
+    }
```

Comment:
This loop might not be needed...
If the DAG does not use the input column, it is not removed.

Reply:
It's uneeded indeed, thanks. Removed it